rust/src/countme.rs

```diff
@@ -0,0 +1,131 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+use anyhow::{bail, Result};
+use curl::easy::Easy;
+use os_release::OsRelease;
+use std::path;
+
+mod cookie;
+mod repo;
+
+// Default variant name used in User Agent
+const DEFAULT_VARIANT_ID: &str = "unknown";
+
+// Send a GET request to 'url' with 'ua' as User Agent and discard the response
+fn send_countme(url: &str, ua: &str) -> Result<()> {
+    println!("Sending request to: {}", url);
+    let mut handle = Easy::new();
+    handle.url(&url)?;
+    handle.useragent(&ua)?;
+    {
+        let mut transfer = handle.transfer();
+        transfer.write_function(|new_data| Ok(new_data.len()))?;
+        transfer.perform()?;
+    }
+    Ok(())
+}
+
+/// Main entrypoint for countme
+fn countme_main(_args: &[String]) -> Result<()> {
+    // Silently skip if we are not run on an ostree booted system
+    if !path::Path::new("/run/ostree-booted").exists() {
+        println!("Skipping: Not running on an ostree based system.");
+        return Ok(());
+    }
+
+    // Load repo configs and keep only those where countme is enabled
+    let mut repos: Vec<repo::Repo> = self::repo::all()?;
+    repos.retain(|r| r.count_me());
+
+    // Load timestamp cookie
+    let cookie = cookie::Cookie::new();
+    // Skip this run if we are not in a new counting window
+    if cookie.existing_window() {
+        println!("Skipping: Not in a new counting window.");
+        return Ok(());
+    }
+
+    // Read /etc/os-release
+    let release: OsRelease = OsRelease::new()?;
+    let variant = match release.extra.get("VARIANT_ID") {
+        Some(s) => s,
+        None => DEFAULT_VARIANT_ID,
+    };
+
+    // Setup User Agent. The format is:
+    // libdnf (NAME VERSION_ID; VARIANT_ID; OS.BASEARCH)
+    // libdnf (Fedora 31; server; Linux.x86_64)
+    // See `user_agent` option in:
+    // https://dnf.readthedocs.io/en/latest/conf_ref.html?highlight=user_agent#options-for-both-main-and-repo
+    let ua = format!(
+        "rpm-ostree ({} {}; {}; {}.{})",
+        release.name,
+        release.version_id,
+        variant,
+        "Linux",
+        std::env::consts::ARCH
+    );
+    println!("Using User Agent: {}", ua);
+
+    // Compute the value to send as window counter
+    let counter = cookie.get_window_counter();
+
+    // Send Get requests, track successfully ones and do not exit on failures
+    let mut successful_requests: u64 = 0;
+    for r in &repos {
+        let url = format!("{}&countme={}", &r.metalink(&release.version_id), counter);
+        match send_countme(&url, &ua) {
+            Ok(()) => successful_requests += 1,
+            Err(e) => println!("Request '{}' failed: {}", url, e),
```

Comment:
I think it'd be cleaner to do something like this:
```
let successful = repos.iter().fold(0, |acc, r| {
    let url = format!("{}&countme={}", &r.metalink(&release.version_id), counter);
        match send_countme(&url, &ua) {
            Ok(_) => acc + 1,
            Err(e) => {
                eprintln!("Request '{}' failed: {}", url, e);
                acc
            }
        }
    }
});
```

Note use of `eprintln!` to print to stderr.

Reply:
(Totally optional though)